pylinalg/func/vector.py

```diff
@@ -116,3 +116,195 @@ def vector_apply_matrix(vectors, matrix, /, *, w=1, out=None, dtype=None):
     if dtype is not None:
         out = out.astype(dtype, copy=False)
     return out
+
+
+def vector_unproject(vector, projection_matrix, /, *, depth=0, out=None, dtype=None):
+    """
+    Un-project a vector from 2D space to 3D space.
+
+    Find a ``vectorB`` in 3D euclidean space such that the projection
+    ``projection_matrix @ vectorB`` yields the provided vector (in 2D euclidean
+    space). Since the solution to the above is a 1D subspace of 3D space (a
+    line), ``depth`` is used to select a single vector within.
+
+    Parameters
+    ----------
+    vector : ndarray, [2]
+        The vector to be un-projected.
+    projection_matrix: ndarray, [4, 4]
+        The camera's intrinsic matrix.
+    depth : number, optional
+        The distance of the unprojected vector from the camera.
```

Comment:
Is this the distance in world units, or a ratio of the distance between far and near? I'd personally prefer the latter, but either way, I'd like that to be expressed in the docstring

Reply:
I might be wrong, but I don't think the camera matrix includes information on the distances of the far and near planes. We could add these as additional kwargs, but it might be cleaner to ask for distance in camera-space units (which may be different from world units if there is scaling between both).

Reply:
I'm having trouble explaining this well. Let me try a different angle. The typical use case for this function is to implement mouse picking, as done here for example: https://gamedev.stackexchange.com/a/12370/31164

Reply:
Ah. The code you linked takes the screen coordinates of the mouse, normalizes them, then picks the corresponding _two_ points on the near and the far plane respectively, unprojects both, and then takes the difference. With the current signature it would look something like this for us:

```python
# normalized mouse coordinates
v, u = get_screen_pos_of_mouse() / (height, width)

near_camera = unproject((u, v), camera_matrix, depth=near)
far_camera = unproject((u, v), camera_matrix, depth=far)

result = far_camera - near_camera
result = result / np.linalg.norm(result)
```

This seems wasteful though because near_camera, far_camera, and the origin (0, 0, 0) always line up for a perspective camera. You'd only need one `unproject` and you can use any depth you desire. As such, you should also be able to do:

```python
# normalized mouse coordinates
v, u = get_screen_pos_of_mouse() / (height, width)
result = unproject((u, v), camera_matrix, depth=1)
```

(For orthographic cameras the problem is trivial as `unproject(anything, depth=value) = (0, 0, value)`, so users will not use `unproject`.)

The difference between the two approaches is that the length of the vector no longer matches the length of the viewing frustum measured along that ray; however, the code normalizes this vector anyway so that's likely unimportant :)

Btw, if the main use case is to cast a ray from the camera into the scene, we could add a `depth=None` default which will return a unit length vector, which saves the normalization step.

Reply:
> Is this the distance in world units, or a ratio of the distance between far and near? I'd personally prefer the latter, but either way, I'd like the unit to be expressed in the docstring for the depth parameter

And to answer the initial question here, the distance would be in world units.

Reply:
@Korijn As far as I can see, this is the only conversation that needs resolving before this PR is ready. Are you okay with depth being in local coordinates instead of being a ratio (where 0 means on the near plane and 1 on the far plane)?

Reply:
Yes!